### ISSUE TITLE
Remove legacy code from twisted.internet._signals, as part of port to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5986]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5986|
|Type|enhancement|
|Created|2012-09-14 17:58:04Z|
|Branch|https://github.com/twisted/twisted/tree/signals-5986-3|

`twisted.internet._signals` and its attendant C module are mostly devoted to versions of Python before 2.6. The unnecessary code can be removed, and the remaining code ported to Python 3.

<details><summary>Searchable metadata</summary>

```
trac-id__5986 5986
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_signals_5986_3 branches/signals-5986-3
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1347645484000000 1347645484000000
changetime__1348858392000000 1348858392000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [35678]) Branching to 'signals-5986'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

This depends on [#5974](https://github.com/twisted/twisted/issues/5974).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

(In [35723]) Branching to 'signals-5986-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst set owner to @exarkun|
|-|-|

OK, some code removed, and it works on Python 3. I also filed [#5996](https://github.com/twisted/twisted/issues/5996). I've started a buildbot run:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/signals-5986-2

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @itamarst|
|-|-|

There's some Windows failures.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

I am blocked on [#4286](https://github.com/twisted/twisted/issues/4286) - if that's still a valid ticket, I need to unkill the _sigchld.c version in the code.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst set owner to @exarkun|
|-|-|

New build looks happy on Windows: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/signals-5986-2

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @itamarst|
|-|-|

  1. Would be nice to unify the signal imports in `twisted/internet/_signals.py` now that no tricky `ImportError` handling is required.
  1. Considering the existence of alternate Python runtimes like PyPy and even other platforms like BSD (with kqueue support for receiving signal notifications), and the highly intricate nature of signal handling itself, it seems likely that at some point in the future we might want to re-introduce multiple approaches, with the best selected at runtime.  In this case, we'll probably want to apply the unit tests to each implementation (somewhat) automatically.  Preserving the mixin-based approach to defining the tests would probably make it more obvious to a future developer how to do this, and would have been less work on this branch too.  Of course, the work is done now and it is only speculative that we'll want to use the now-deleted testing infrastructure in the future.  So I guess this review comment should read "Oh well".
  1. The skip of `test_openFileDescriptors` would be better done using the skip attribute, rather than raising `SkipTest`.  Pre-existing defect, so just file a new ticket if you'd prefer.
  1. `test_systemCallUninterruptedByChildExit` was not previously skipped on all non-POSIX platforms.  Why is it skipped on all non-POSIX platforms now? ''(As a side note, whoever allowed `test_systemCallUninterruptedByChildExit` to be added should be severely punished.)''
  1. Minor conflict in `admin/_twistedpython3.py`
  1. And I'm kinda conflicted about the .removal file.  We already announced we're dropping Python 2.5 support.  It seems unnecessary to announce each individual change where this manifests.  Is someone going to read the news file and care about this?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst commented|
|-|-|

(In [35823]) Branching to 'signals-5986-3'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst commented|
|-|-|

(In [35825]) Address review comments. Refs [#5986](https://github.com/twisted/twisted/issues/5986)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst set owner to @exarkun|
|-|-|

  1. Done.
  1. Fair point, I'll remember next time I do this sort of thing.
  1. Done.
  1. SIGCHLD is only an issue on POSIX, which is what this testing. I removed the check, though, we'll see if it's happy in buildbot.
  1. Done.
  1. True. Switched to misc.

Started another run: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/signals-5986-3

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @itamarst|
|-|-|

Great.  Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst set status to `closed`|
|-|-|

(In [35835]) Merge signals-5986-3.

Author: itamar
Review: exarkun
Fixes: [#5986](https://github.com/twisted/twisted/issues/5986), [#4286](https://github.com/twisted/twisted/issues/4286)

Port twisted.internet._signals to Python 3, dropping unnecessary pre-2.6 code along the way.
